### PR TITLE
Update SDK Development Doc

### PR DESCRIPTION
**Pull request by lpabon**
**What this PR does / why we need it**:
Update SDK Development documentation

### CODE REVIEW

File: docs/dev-sdk.md [hunk 1]
```diff
@@ -1,59 +1,199 @@
...
+    * Here is an example for [OpenStorageClusterService](https://libopenstorage.github.io/w/generated-api.html#serviceopenstorageapiopenstoragecluster)
+
+##### SDK APIs
+
+* If it is a new service, then it should have `Create`, `Inspect`, `Delete`, or `Enumerate` style APIs, if possible.
+* All APIs **must** have a single message for the request and a single message for the response with the following style: `Sdk<Service Type><Api Name>Request|Response`
+
+##### Enums
+
+* Following the [Google protobuf style for enums](https://developers.google.com/protocol-buffers/docs/style#enums)
```
Comment (lpabon) [inline docs/dev-sdk.md:46]: `Follow` not `Following`

File: docs/dev-sdk.md [hunk 2]
```diff
@@ -1,59 +1,199 @@
...
-    * `string` values are for ids or strings. They are not meant to marshal information as a `yaml`, create a concrete _message_ instead.
+#### Version
+
+Any changes to the protocol must bump the version by one. On the _master_ branch, the _minor_ number is bumped. On _release_ branches, the _patch_ number is bumped.
+
+#### Readability 
+
+**APIs Must be readable**. SDK APIs and values must be concrete clear values. They are used not just by Portworx, but also non-Portworx developers which do not have an understanding of the internals of the Portworx cluster.
+
+#### `string` values.
```
Comment (ggriffiths) [inline docs/dev-sdk.md:26]: maybe include a note on avoid opaque map[string]string options when possible? I remember you mentioning this in PR reviews.

File: docs/dev-sdk.md [hunk 3]
```diff
@@ -1,59 +1,199 @@
...
-* Messages
-    * Try not to use `uint64`. Instead try to use signed `int64`. (There is a reason for this which is why CSI changed all uint64s to in64s in version 0.2, but I can find out why.)
-* Documentation
-    * It is imperative that the documentation is correct since it is used to automatically generate the documentation for https://libopenstorage.github.io .The documentation for these values in the proto files can be in Markdown format.
-
-**NOTE** Most importantly is that these APIs _must_ be supported forever. They will almost never be deprecated since at some point we will have many versions of the clients. So please be clear and careful on the API you create.
+    * Instead of passing "Done", or "paused", use enums for these value, making it clear to the reader.
 
-## Creating new API
+#### SDK Services
```
Comment (ggriffiths) [inline docs/dev-sdk.md:32]: Can include a note about updating default users (system.user) with any services that make sense to add when creating a new service.

**Review by ggriffiths — APPROVED** [review]
Mostly looks good, just a few suggestions

File: docs/dev-sdk.md [hunk 4]
```diff
@@ -1,59 +1,199 @@
...
+```proto
+message Foo {
+  reserved 6;
+}
+```
+
+It is essential that no values override the field number when updating or replacing. From the guide:
+
+> Note: If you update a message type by entirely removing a field, or commenting it out, future users can reuse the field number when making their own updates to the type. This can cause severe issues if they later load old versions of the same .proto, including data corruption, privacy bugs, and so on.
+
```
Comment (adityadani) [inline docs/dev-sdk.md:163]: With regards to field numbers, in a scenario where there is a oneof or sub message you could use the following pattern 

```
  string s3_storage_class = 7;
  // Start at field number 200 to allow for expansion
  oneof credential_type {
    // Credentials for AWS/S3
    SdkAwsCredentialRequest aws_credential = 200;
    // Credentials for Azure
    SdkAzureCredentialRequest azure_credential = 201;
    // Credentials for Google
    SdkGoogleCredentialRequest google_credential = 202;
  }
```

**Review by adityadani — APPROVED** [review]

File: docs/dev-sdk.md (hunk 4 above)
Comment (lpabon) [reply docs/dev-sdk.md:163]: Nice I will add that

**Pull request merged**